docs/standard/serialization/snippets/system-text-json-how-to/csharp/RoundtripDataTable.cs

```diff
@@ -0,0 +1,246 @@
+                                jsonWriter.WriteNumber(key, doubleValue);
+                            }
+                            break;
+                        default:
+                            jsonWriter.WriteString(key, valueString);
+                            break;
+                    }
+                }
+                jsonWriter.WriteEndObject();
+            }
+            jsonWriter.WriteEndArray();
+        }
+    }
+    public static class Extensions
+    {
+        public static DataTable JsonElementToDataTable(this JsonElement dataRoot)
+        {
+            var dataTable = new DataTable();
+            var firstPass = true;
+            foreach (var element in dataRoot.EnumerateArray())
+            {
+                if (firstPass)
+                {
+                    foreach (var col in element.EnumerateObject())
+                    {
+                        var colValue = col.Value;
+                        dataTable.Columns.Add(new DataColumn(col.Name, colValue.ValueKind.ValueKindToType(colValue.ToString())));
+                    }
+                    firstPass = false;
+                }
+                var row = dataTable.NewRow();
+                foreach (var col in element.EnumerateObject())
+                {
+                    row[col.Name] = col.Value.JsonElementToTypedValue();
+                }
+                dataTable.Rows.Add(row);
+            }
+            return dataTable;
+        }
+
+        private static Type ValueKindToType(this JsonValueKind valueKind, string value)
+        {
+            switch (valueKind)
+            {
+                case JsonValueKind.String:      // 3
+                    return typeof(System.String);
+                case JsonValueKind.Number:      // 4    
+                    if (Int64.TryParse(value, out var intValue))
+                    {
+                        return typeof(System.Int64);
+                    }
+                    else
+                    {
+                        return typeof(System.Double);
+                    }
+                case JsonValueKind.True:        // 5
+                case JsonValueKind.False:       // 6
+                    return typeof(System.Boolean);
+                case JsonValueKind.Undefined:   // 0
+                    return null;
+                case JsonValueKind.Object:      // 1 
+                    return typeof(System.Object);
+                case JsonValueKind.Array:       // 2
+                    return typeof(System.Array);
+                case JsonValueKind.Null:        // 7
+                    return null;
+                default:
+                    return typeof(System.Object);
+            }
+        }
+
+        private static object JsonElementToTypedValue(this JsonElement jsonElement)
+        {
+            switch (jsonElement.ValueKind)
+            {
+                case JsonValueKind.Object:      // 1  (these need special handling)?
```

Comment:
nit: separate these cases since they have different behavior.

Reply:
>separate these cases since they have different behavior.

and throw NotSupportedException for them?